### ISSUE TITLE
Ansible 2.x: local connector misbehaves when using with_items

### ISSUE THREAD

**Issue by ThomasSteinbach** [ansible/ansible#13624]
Ansible crashes when creating links with_items and using `ansible_connection=local`. Here an example playbook for reproducing the behavior:

```
- name: the last task will throw an error
  hosts: localhost
  gather_facts: no

  tasks:
    - name: create files - that will work
      copy:
        content: "test"
        dest: "/tmp/{{ item }}"
      with_items:
        - testfile1
        - testfile2

    - name: create links - that will crash
      become: yes
      file:
        src: "/tmp/{{ item }}"
        dest: "/tmp/{{ item }}link"
        state: link
      with_items:
        - testfile1
        - testfile2
```

The output is:

```
PLAY [the last task will throw an error] ***********************************

TASK [create files - that will work] *************************************************************
ok: [localhost] => (item=testfile1)
ok: [localhost] => (item=testfile2)

TASK [create links - that will crash] *************************************************************
fatal: [localhost]: FAILED! => {"failed": true, "msg": "ERROR! privilege output closed while waiting for password prompt:\nBECOME-SUCCESS-smnurvzoeqwqcapbnmmyrhlaautxdumn\n{\"src\": \"/tmp/testfile2\", \"group\": \"root\", \"uid\": 0, \"dest\": \"/tmp/testfile2link\", \"changed\": false, \"owner\": \"root\", \"state\": \"link\", \"gid\": 0, \"mode\": \"0777\", \"invocation\": {\"directory_mode\": null, \"force\": false, \"remote_src\": null, \"path\": \"/tmp/testfile2link\", \"owner\": null, \"follow\": false, \"group\": null, \"state\": \"link\", \"content\": null, \"serole\": null, \"diff_peek\": null, \"setype\": null, \"dest\": \"/tmp/testfile2link\", \"selevel\": null, \"original_basename\": null, \"regexp\": null, \"validate\": null, \"src\": \"/tmp/testfile2\", \"seuser\": null, \"recurse\": false, \"delimiter\": null, \"mode\": null, \"backup\": null}, \"size\": 14}\n"}

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1 
```

I think the local connector misbehaves with more then just the _file_ module, when using `with_items`. I have another issue open ([pip module does not work with_items in Ansible 2.0 #2552](https://github.com/ansible/ansible-modules-core/issues/2552)) where i suspected the _pip_ module having problems `with_items`. However, the general error remains the same in both issues:

```
ERROR! privilege output closed while waiting for password prompt
```

During the investigation of the other issue, I found out that the error for it was introduced with commit https://github.com/ansible/ansible/commit/53e86f3130145ce40606cec32aef2858756a20ab, which was merged with PR https://github.com/ansible/ansible/pull/13261. I haven't checked if it is the same for this issue, but it could be a first trace.

**Comment by bcoca** [comment]
note that doing a local action that affects files in a host loop can create race conditions, you basically have the same forks that normally go do different machines trying to update the same files on the same machine. I recommend using `run_once: true` in these cases.

**Comment by ThomasSteinbach** [comment]
Mhh, maybe I did not get exactly what you mean. When I have a look on my example playbook above, I didn't see that there is one file touched twice at the same time. The both tasks were executed consecutively and within the `with_items` loop different files where referenced.

But maybe I forget to mention a detail in my bug description above. Does it change your point of view, when I say that the same playbook works, when setting

```
localhost ansible_ssh_host=127.0.0.1
```

in the inventory instead of `localhost ansible_connection=local`? Not even that the file module then works like expected, also the pip module works when using a local ssh connection, as stated in issue https://github.com/ansible/ansible-modules-core/issues/2552.

As the modules should behave the same for every `ansible_connection` type, I think there is a general faux pas within the local connector logic.

**Comment by Yannig** [comment]
I can confirm. Using local connection with sudo, the playbook above crash (if files did not exist)

**Comment by ThomasSteinbach** [comment]
Thank you @Yannig for the detail. After that much investigation I become blind for the details. If you look at the error thrown, it looks like `ansible_connection=local` has a problem `with_items` in combination with `become`. And maybe the bug then is independent of the module (file and pip in this case), but that I haven't examined.

**Comment by Yannig** [comment]
If sudo context is iniated, Ansible is working fine. If forcing sudo password (with --ask-become-pass) on already sudoed context, it crash. I think, it have something to deal with this.

**Comment by Yannig** [comment]
A very ugly workaround that is doing the trick:

``` patch
diff --git a/lib/ansible/plugins/connection/local.py b/lib/ansible/plugins/connection/local.py
index e69281d..166fd8f 100644
--- a/lib/ansible/plugins/connection/local.py
+++ b/lib/ansible/plugins/connection/local.py
@@ -36,6 +36,7 @@ except ImportError:
     from ansible.utils.display import Display
     display = Display()

+already_sudoed = False

 class Connection(ConnectionBase):
     ''' Local based connections '''
@@ -59,6 +60,7 @@ class Connection(ConnectionBase):
         return self

     def exec_command(self, cmd, in_data=None, sudoable=True):
+        global already_sudoed
         ''' run a command on the local host '''

         super(Connection, self).exec_command(cmd, in_data=in_data, sudoable=sudoable)
@@ -82,7 +84,8 @@ class Connection(ConnectionBase):
         )
         display.debug("done running command with Popen()")

-        if self._play_context.prompt and sudoable:
+        if not already_sudoed and self._play_context.prompt and sudoable:
+            already_sudoed = True
             fcntl.fcntl(p.stdout, fcntl.F_SETFL, fcntl.fcntl(p.stdout, fcntl.F_GETFL) | os.O_NONBLOCK)
             fcntl.fcntl(p.stderr, fcntl.F_SETFL, fcntl.fcntl(p.stderr, fcntl.F_GETFL) | os.O_NONBLOCK)
             become_output = ''
```

Basicaly, we are sudoing the first time but not next run as we have already inialized sudo context.

**Comment by Yannig** [comment]
@ThomasSteinbach any feedback about this change?

**Comment by bcoca** [comment]
so running this i get:
```

**Comment by bcoca** [comment]
when running with ssh as a connection i get the error in the clear:
```
failed: [localhost] => (item=testfile) => {"failed": true, "item": "testfile", "msg": "src file does not exist, use \"force=yes\" if you really want to create the link: /tmp/testfile", "path": "/tmp/testfilelink", "src": "/tmp/testfile", "state": "absent"}```

**Comment by bcoca** [comment]
ignore previous messages, was bad copy paste, had testfile vs testfile2

**Comment by kastixx** [comment]
I don't know if this is related, but I've run into a similar problem involving with_items and local connection.

This commit: kastixx/ansible@6a7d07041fb87dd8ad7f1a5df17fceb5a1f17f64 seems to fix this issue for me.

**Comment by stefanhorning** [comment]
I have the same problem with a loop and a local connection. become / sudo only works for the first iteration and starts failing from the second. In my case it was about linking some files. So when I do it on a remote host it worked and when I do the same actions in single tasks they all work. So I also narrowed it down to sudo only working for the first iteration when the connection is local.

**Comment by stefanhorning** [comment]
I could solve it though by removing the --sudo from the command line options and usign `become: true` in the playbook instead.

**Comment by stefanhorning** [comment]
Diff between red and green build:
https://github.com/mediapeers/ansible-role-ruby/compare/3255d5406f05d7a684f983c8bfb780b245e43ee1...mediapeers:c9d67fc0ffab7151f37f6d5c1fe877ea7fc7ad31

This was the error before:
https://travis-ci.org/mediapeers/ansible-role-ruby/builds/102382742#L234

In case that helps anybody...

**Comment by andrewtchin** [comment]
I'm seeing the same issues as @Yannig
`ansible-playbook -vvv playbooks/common.yml --ask-become-pass -c local` fails with the same error, but `sudo ansible-playbook -vvv playbooks/common.yml -c local` works fine. 
I tested the patch from 22 December and it worked. Want to make a PR @Yannig?

**Comment by andrewtchin** [comment]
Also tested @kastixx's patch, works.

**Comment by Yannig** [comment]
@andrewtchin can you provide more details about your bug (playbook and os)?
This bug have been fix by https://github.com/ansible/ansible/pull/13793
Can you give a try with last devel version of Ansible?

**Comment by andrewtchin** [comment]
@Yannig good call devel worked thanks!

**Comment by abadger** [comment]
This issue appears to have been fixed by #13793.  I'm going to go ahead and close it.
